Спринт 9

NastasyaRosh opened the pull request.

NastasyaRosh committed:
Спринт 9. Попытка 1.

sergeymamonov commented on src/test/java/ru/yandex/practicum/filmorate/TestFilmController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Test
+    public void errorValidation() {
+        Film withFalseDate = Film.builder().name("Фильм").description("Ложная дата.").releaseDate(LocalDate.of(1895, 12, 27)).duration(90).build();
```
Обратите внимание, что в каждом тесте создается фильм. То есть можно вынести создание фильма в метод beforeEach() и там создавать корректный филь, а в каждом тесте менять одно из полей у уже созданного фильма на некорректное значение.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+@Slf4j
+public class FilmController {
+    final LocalDate firstFilm = LocalDate.of(1895, 12, 28);
+    private HashMap<Integer, Film> films = new HashMap<>();
```
Нужно стараться определять все на уровне интерфейсов. Здесь это можно сделать изменив на Map<Integer, Film> films

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+@RequestMapping("/films")
+@Slf4j
+public class FilmController {
+    final LocalDate firstFilm = LocalDate.of(1895, 12, 28);
```
Эту переменную можно определить как константу, дав ей соответствующие модификаторы и написание имени с помощью заглавных букв.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @PostMapping
+    public Film addFilm(@Valid @RequestBody Film film) {
+            if (film.getReleaseDate().isBefore(firstFilm)) {
```
Лучше выделить валидацию в отдельный метод. И тогде будет лаконичнее
addFilm(@Valid @RequestBody Film film) {
validate(film);
films.put(setId(film), film);
...

Else не нужен, так как если будет ошибка, то код дальше throw не пойдет, если ошибка нигде не обрабатывается.

Для остальных методов аналогично.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,56 @@
+@Slf4j
+public class UserController {
+
+    private HashMap<Integer, User> users = new HashMap<>();
```
Классы для юзера аналогичны классам для фильма, поэтому и замечания аналогичные. Изменить на интерфейс.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,61 @@
+import ru.yandex.practicum.filmorate.model.Film;
+
+import javax.validation.Valid;
+import javax.validation.ValidationException;
```
Какой класс предполагается использовать? Который импортируется или созданные в пакете exception ValidationException? Если импортируемый, то созданный класс ValidationException необходимо удалить.

NastasyaRosh replied:
Решила использовать импортируемый, а свой удалить

NastasyaRosh committed:
Спринт 9. Попытка 2.

NastasyaRosh committed:
-refactor обобщила hashmap до map в обоих контроллерах
-refactor сделала переменную firstFilm константой
-feat вынесла валидацию и проверку на существование в отдельные методы
-delete решила удалить свой ValidationException и использовать импортируемый
-refactor в тестах вынесла создание базового объекта в before each
-style поудаляла лишние пустые строки и табы

NastasyaRosh commented:
Внесла изменения и отправила новый вариант.

NastasyaRosh committed:
Merge remote-tracking branch 'origin/controllers-films-users' into controllers-films-users

NastasyaRosh commented:
Внесла изменения и отправила новый вариант.
Правда комментарий на яндекс практикуме заметила после того как запушила, поэтому дозалила еще раз с нормальным комментарием к коммиту о том, что сделано :)

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @GetMapping
+    public Collection<Film> allFilms() {
```
Лучше сначала размешать в классе публичные методы, а затем приватные.

sergeymamonov replied:
Лучше возвращать List. Collection - слишком общий интерфейс, в нем может быть и массив, и множестве и так далее.

NastasyaRosh committed:
Дополнительные правки спринта 9
-refactor в методе GET allFims и allUsers зменила возвращаемое значение с collections на list
-style изменила порядок следования методов - сначала публичные, затем приватные

NastasyaRosh committed:
Дополнительные правки спринта 9
-fix в методе GET allFims и allUsers исправила ошибку конвертации hashMap.values (Collections) в ArrayList
